discovery_engine/lib/src/infrastructure/repository/hive_source_preference_repo.dart

```diff
@@ -0,0 +1,53 @@
+// Copyright 2022 Xayn AG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import 'package:hive/hive.dart' show Hive, Box;
+import 'package:xayn_discovery_engine/discovery_engine.dart' show Source;
+import 'package:xayn_discovery_engine/src/domain/models/source_preference.dart'
+    show SourcePreference, PreferenceMode;
+import 'package:xayn_discovery_engine/src/domain/repository/source_preference_repo.dart'
+    show SourcePreferenceRepository;
+import 'package:xayn_discovery_engine/src/infrastructure/box_name.dart'
+    show sourcePreferenceBox;
+
+class HiveSourcePreferenceRepository implements SourcePreferenceRepository {
+  Box<SourcePreference> get box =>
+      Hive.box<SourcePreference>(sourcePreferenceBox);
+
+  @override
+  Future<Set<Source>> getFavorites() async {
+    return box.values
+        .where((el) => el.mode == PreferenceMode.favorite)
+        .map((el) => el.source)
+        .toSet();
+  }
+
+  @override
+  Future<Set<Source>> getExcluded() async {
+    return box.values
+        .where((el) => el.mode == PreferenceMode.excluded)
+        .map((el) => el.source)
+        .toSet();
+  }
+
+  @override
+  Future<void> save(SourcePreference filter) async {
+    await box.put(filter.source.toString(), filter);
```

Comment:
`Source` should have a `value` getter which returns the underlying `String`, it's better to use that then `toString()` representation. `toString()` could change and is often used for debugging.
```suggestion
    await box.put(filter.source.value, filter);
```